Ticket2705/2704 journal viewer filtering and sorting

adamb56789 opened the pull request:
### Description of work

Added GUI elements for controlling the search. Added functionality to the back-end to get inputs from the GUI and construct the more complicated SQL queries required. Added new unit tests for the journal back-end. Added ability to sort by columns.

### Ticket

Filtering: https://github.com/ISISComputingGroup/IBEX/issues/2705
Sorting: https://github.com/ISISComputingGroup/IBEX/issues/2704

### Acceptance criteria

- [x] You can filter entries by Run Number: display entries with run number between a minimum value and a maximum value.
- [x] You can filter entries by RB Number: display entries associated with a single RB number.
- [x] You can filter entries by Title: display entries containing specified text.
- [x] You can filter entries by Users: display entries containing specified text.
- [x] You can filter entries by Start Time: display entries with Start Times between a minimum date and a maximum date.
- [x] The column sorting behaves as specified in the ticket.

- [x] All of the buttons behave as a user would expect.
- [x] The GUI design is clear, easy to use, and fits the correct style.
- [x] Added units tests are comprehensive and pass, and old unit tests still pass.

### Documentation
Small changes to:
https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/The-Journal-Viewer
https://github.com/ISISComputingGroup/ibex_user_manual/wiki/Views

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

Tom-Willemsen commented on base/uk.ac.stfc.isis.ibex.journal/src/uk/ac/stfc/isis/ibex/journal/JournalSqlStatement.java:
```diff
@@ -0,0 +1,165 @@
+     * @param sortField The field to sort with.
+     */
+    public void addAscendingSort(JournalField sortField) {
+        sorts.add(new JournalSort(sortField, JournalSortDirection.ASCENDING));
```
This method appears to be unused from anywhere except test code. It can be removed.

adamb56789 replied:
Reworked how sorts are added so there is only one `addSort` method now, which is in `JournalSearchParameters`.

Tom-Willemsen commented on base/uk.ac.stfc.isis.ibex.journal/src/uk/ac/stfc/isis/ibex/journal/JournalParameters.java:
```diff
@@ -0,0 +1,134 @@
+/**
+ * Encapsulates journal search parameters.
+ */
+public class JournalParameters {
```
If the javadoc is correct, then a better name for this class might be `JournalSearchParameters`, which makes it explicit that these are just the parameters for a search, not for the whole journal viewer.

adamb56789 replied:
Changed it to `JournalSearchParameters`.

Tom-Willemsen commented on base/uk.ac.stfc.isis.ibex.journal.tests/src/uk/ac/stfc/isis/ibex/journal/tests/JournalSqlStatementTest.java:
```diff
@@ -0,0 +1,115 @@
+        String t = statement.createWhereTemplate();
+        
+        assertEquals(t, " WHERE run_number >= ? AND run_number <= ?");
+    }
```
We should have a test for the dates as well as intger and string search types

adamb56789 replied:
Added one.

Tom-Willemsen commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/SearchInput.java:
```diff
@@ -0,0 +1,309 @@
+                    }
+                });
+            }
+        });
```
This method is very long. UI methods are allowed to be a bit longer than logic methods, but this method is still a bit excessive. Is there a way to split it up?

adamb56789 replied:
I've moved the listeners and bindings out of the method and got rid of some duplication, which shortened it from 145 lines to 90 lines.

Tom-Willemsen commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/SearchInput.java:
```diff
@@ -0,0 +1,309 @@
+        } else if (stackSearch.topControl == cmpUsers) {
+            return Optional.of(txtSearchUsers.getText());
+        } else {
+            return null;
```
This should presumably be `Optional.empty()` rather than null - otherwise we lose all the benefits of using an `Optional`

adamb56789 replied:
Changed it to `Optional.empty()`.

Tom-Willemsen commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/SearchInput.java:
```diff
@@ -0,0 +1,309 @@
+    /**
+     * Clears all of the user input and resets the buttons.
+     */
+    public void clearInput() {
```
This logic doesn't belong in the UI layer.

What we want to do is:
- Databind all of the UI items to a viewmodel
- The `clear` operation is implemented in the viewmodel, which updates things back to their defaults
- The databinding then takes care of updating the UI.

adamb56789 replied:
After discussion we decided not to do this; the code is still maintainable the way it is without using bindings.

adamb56789 commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java:
```diff
@@ -140,80 +185,167 @@ protected void addColumns() {
+
+        Thread searchJobThread = new Thread(searchJob);
+
+        searchJobThread.start();
```
Decided not to do this since it is done like this for existing searches already.

adamb56789 replied:
Reworked the way database fetch jobs are handled (partially for other reasons) and they now submit to a threadpool in the model.

adamb56789 commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java:
```diff
@@ -140,80 +185,167 @@ protected void addColumns() {
+                    parameters.setTimes(searchInput.getStartTimeFrom(), searchInput.getStartTimeTo());
+                } else {
+                    parameters.setSearchString(searchInput.getActiveSearchText());
+                }
```
We started this refactor on branch ```start_journal_search_refactor```, but it was getting too hard so we decided not to continue. Might come back to it later.

adamb56789 commented on base/uk.ac.stfc.isis.ibex.journal.tests/src/uk/ac/stfc/isis/ibex/journal/tests/JournalSqlStatementTest.java:
```diff
@@ -0,0 +1,115 @@
+    public void test_GIVEN_no_parameters_WHEN_create_where_template_THEN_nothing_returned() {
+        JournalSqlStatement statement = new JournalSqlStatement(parameters, PAGE_NUMBER, PAGE_SIZE, connection);
+        
+        String t = statement.createWhereTemplate();
```
Changed from `t` to `template`.

adamb56789 commented on base/uk.ac.stfc.isis.ibex.journal/src/uk/ac/stfc/isis/ibex/journal/JournalParameters.java:
```diff
@@ -0,0 +1,134 @@
+    private Optional<Calendar> fromTime = Optional.empty();
+    private Optional<Calendar> toTime = Optional.empty();
+    
+    private boolean set = false;
```
Added a javadoc.

adamb56789 commented on base/uk.ac.stfc.isis.ibex.journal/src/uk/ac/stfc/isis/ibex/journal/JournalSqlStatement.java:
```diff
@@ -0,0 +1,165 @@
+    private int pageNumber;
+    private int pageSize;
+    private Connection connection;
+    private ArrayList<JournalSort> sorts = new ArrayList<JournalSort>();
```
Implemented sorting as in [#2704](https://github.com/ISISComputingGroup/IBEX/issues/2704)

adamb56789 commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java:
```diff
@@ -59,67 +63,108 @@
+    private DataboundTable<JournalRow> table;
+
+    private static final JournalField[] FIELDS = {JournalField.RUN_NUMBER, JournalField.TITLE, JournalField.START_TIME,
+            JournalField.RB_NUMBER, JournalField.USERS};
```
Used a list instead and moved it to the model.

adamb56789 commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java:
```diff
@@ -140,80 +185,167 @@ protected void addColumns() {
+     * request parameters.
+     */
+    private void search() {
+        if (model != null) {
```
Not sure, got rid of this.

adamb56789 commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java:
```diff
@@ -140,80 +185,167 @@ protected void addColumns() {
+        if (model != null) {
+            int fieldIndex = searchInput.getCmbFilterTypeIndex();
+
+            if (fieldIndex != -1) {
```
Since the combo is initialised to 0, it should be impossible to get it to -1, so removed this check.

adamb56789 commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java:
```diff
@@ -140,80 +185,167 @@ protected void addColumns() {
+            public void run() {
+                progressBar.setVisible(visible);
+            }
+        });
```
Done

DominicOram commented on base/uk.ac.stfc.isis.ibex.journal/src/uk/ac/stfc/isis/ibex/journal/JournalSearchParameters.java:
```diff
@@ -0,0 +1,179 @@
+     * Once some parameters have been set, this variable is set to true
+     * and used to prevent more parameters from being added.
+     */
+    private boolean set = false;
```
If this is behaviour you want to stop other developers from doing you should make it impossible for them to do it at compile time. Otherwise this will lead to exceptions during run time. In this case I think that would be best done by splitting this class into four:
* An abstract base class, whose constructor takes a JournalField
* A SearchString class whose constructor takes a String
* A FromTo class whose constructor takes a fromNumber and toNumber 
* A Time class whose constructor takes a fromTime and toTime

You can also then move the logic of what parts to put in the SQL and what to fill the statement with into these classes too. We can discuss further if this doesn't make sense.

adamb56789 replied:
I've replaced `JournalSearchParameters` and `JournalSqlStatement` with the abstract base class `JournalSearch` and the following subclasses:
- `EmptySearch`
- `SearchString`
- `SearchNumber`
- `SearchTime`

DominicOram replied:
This is much improved. Even if it is a bit more verbose it's made it far more extensible.

DominicOram commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/JournalViewerView.java:
```diff
@@ -140,80 +196,144 @@ protected void addColumns() {
+            }
+        });
+
+        model.addPropertyChangeListener("runs", new PropertyChangeListener() {
```
You can get rid of some boiler plate here:
```
        model.addPropertyChangeListener("runs", e -> 
                DISPLAY.asyncExec(() -> {
                        setProgressIndicatorsVisible(true);
                        changeTableColumns();
                        table.setRows(model.getRuns());
                        setProgressIndicatorsVisible(false);
                    }));
```

adamb56789 replied:
Done

DominicOram commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/SearchInput.java:
```diff
@@ -0,0 +1,282 @@
+        cmbFilterType.addSelectionListener(new SelectionAdapter() {
+            @Override
+            public void widgetSelected(SelectionEvent e) {
+                DISPLAY.asyncExec(new Runnable() {
```
Can remove some boilerplate:
```
                DISPLAY.asyncExec(() -> {
                        stackSearch.topControl = cmpFilters.get(cmbFilterType.getSelectionIndex());
                        cmpSearch.layout();
                    });
```

adamb56789 replied:
Done

adamb56789 commented on base/uk.ac.stfc.isis.ibex.ui.journalviewer/src/uk/ac/stfc/isis/ibex/ui/journalviewer/models/JournalViewModel.java:
```diff
@@ -92,13 +93,37 @@ public String getLastUpdate() {
+     * @param parameters The parameters to search with.
+     * @return a CompletableFuture
+     */
+    public CompletableFuture<Void> search(JournalSearchParameters parameters) {
```
It isn't, removed it.

adamb56789 commented:
- Created the ticket [#4550](https://github.com/ISISComputingGroup/IBEX/issues/4550)
- Added press enter to search when in any of the search parameter entry boxes
- Added the same thing to the log message search
- Added the message "No results found with current search parameters"
- Increased the size of the run number search boxes so they can fit 8 digits
- Added an indicator when a column is sorted
- Searching or clearing no longer resets the sort to default
